src/bsoncxx/lib/bsoncxx/v1/types/view.cpp

```diff
@@ -34,6 +53,199 @@ BSONCXX_V1_TYPES_XMACRO(X)
 static_assert(is_semitrivial<view>::value, "bsoncxx::v1::types::view must be semitrivial");
 
+#pragma push_macro("X")
+#undef X
+#define X(_name, _value) constexpr id b_##_name::type_id;
+BSONCXX_V1_TYPES_XMACRO(X)
+#pragma pop_macro("X")
+
+#pragma push_macro("X")
+#undef X
+#define X(_name, _value)                              \
+    b_##_name view::get_##_name() const {             \
+        if (_id != id::k_##_name) {                   \
+            throw v1::exception{code::type_mismatch}; \
+        }                                             \
+        return _b_##_name;                            \
+    }
+
+BSONCXX_V1_TYPES_XMACRO(X)
+#pragma pop_macro("X")
+
+std::error_category const& view::error_category() {
+    class type final : public std::error_category {
+        char const* name() const noexcept override {
+            return "bsoncxx::v1::types::view";
+        }
+
+        std::string message(int v) const noexcept override {
+            switch (static_cast<code>(v)) {
+                case code::zero:
+                    return "zero";
+                case code::type_mismatch:
+                    return "requested type does not match the underlying type";
+                default:
+                    return "unknown: " + std::to_string(v);
+            }
+        }
+
+        bool equivalent(int v, std::error_condition const& ec) const noexcept override {
+            if (ec.category() == v1::source_error_category()) {
+                using condition = v1::source_errc;
+
+                auto const source = static_cast<condition>(ec.value());
+
+                switch (static_cast<code>(v)) {
+                    case code::type_mismatch:
+                        return source == condition::bsoncxx;
+
+                    case code::zero:
+                    default:
+                        return false;
+                }
+            }
+
+            if (ec.category() == v1::type_error_category()) {
+                using condition = v1::type_errc;
+
+                auto const source = static_cast<condition>(ec.value());
+
+                switch (static_cast<code>(v)) {
+                    case code::type_mismatch:
+                        return source == condition::runtime_error;
+
+                    case code::zero:
+                    default:
+                        return false;
+                }
+            }
+
+            return false;
+        }
+    };
+
+    static bsoncxx::immortal<type> const instance;
+
+    return instance.value();
+}
+
+namespace {
+
+v1::stdx::string_view to_sv(char const* data, std::size_t size) {
+    if (!data || size == 0u) {
+        return {};
+    }
+
+    return {data, size};
+}
+
+v1::stdx::string_view to_sv(char const* data) {
+    if (!data) {
+        return {};
+    }
+
+    return to_sv(data, std::strlen(data));
+}
+
+} // namespace
+
+view view::internal::make(bson_value_t const& v) {
```

Comment:
This implements the behavior currently handled by v_noabi's `convert_from_libbson()` in [convert.hh](https://github.com/mongodb/mongo-cxx-driver/blob/ad622035ec9c9ab352b42449d5fcbf426e8782e9/src/bsoncxx/lib/bsoncxx/private/convert.hh), but without exposing `bson_value_t` to dependent components.